src/OptPass.cc

```diff
@@ -60,9 +71,134 @@ bool OptPass::runOnModule(Module&)
 
   outs() << "Got policy file!\n";
 
-  return false;
-}
+	std::map<CallInst*, std::vector<Policy::Direction>> callInstsMap;
+		ErrorOr<std::unique_ptr<PolicyFile>> p = PolicyFile::Open();
+		if (p.getError()) {
+			errs() << p.getError().message() << "\n";
+			errs() << "error when retrieving Policy File \n";
+			return false;
+		}
+		assert(*p);
+		Policy& policy = **p;
 
+    for (Module::iterator i = module.begin(), e = module.end(); i!=e; i++) {
+			for (inst_iterator j = inst_begin(*i), e = inst_end(*i); j != e; ++j) {
+				if (CallInst* callInst = dyn_cast<CallInst>(&*j)) {  
+					auto directions = policy.CallInstrumentation(*callInst->getCalledFunction());
+					if (not directions.empty()) {
+						callInstsMap.insert(std::make_pair(callInst, directions));	
+					}
+				}
+			}
+    }
+		//Iterate over call instructions and create a new instrumented function 
+		//and create a function call to that new function
+		for (std::map<CallInst*, std::vector<Policy::Direction>>::iterator i = callInstsMap.begin(); i != callInstsMap.end(); ++i) {
+			
+			CallInst* callInst = (*i).first;
+			Function* pOldF = callInst->getCalledFunction();
+  		callInst->setCallingConv(pOldF->getCallingConv());
+			std::vector<Value*> argumentValues;
+			for (Use *m = callInst->arg_begin(), *n = callInst->arg_end(); m != n; ++m) {
+				  argumentValues.push_back(m->get());
+			}
+				
+			std::vector<Policy::Direction> directions = (*i).second;
+			for(std::vector<Policy::Direction>::iterator it=directions.begin(); it !=directions.end(); ++it) {
+				Function* pNewF = createInstrFunction(module, (*i).first, *it, policy);
+				IRBuilder<> Builder(&*callInst);
+				CallInst *callToInstr = Builder.CreateCall(pNewF, argumentValues);		
+				callToInstr->setAttributes(pOldF->getAttributes());
+			}
+		}
+    return true;
+}
 
 char OptPass::ID = 0;
 static RegisterPass<OptPass> X("loom", "Loom instrumentation", false, false);
+
+Function* createInstrFunction(Module &module, CallInst* callInst, Policy::Direction direction, Policy& policy) {
```

Comment:
You might also like to take a look at InstrumentationFn.cc to see if there's any overlap there.